### ISSUE TITLE
Inbound SCIM Support for Users

### ISSUE THREAD

**Issue by hifabienne** [zitadel/zitadel#8140]
Inbound SCIM allows external identity providers to provision and manage user accounts within Zitadel according to the SCIM Standard. 

This feature would enable:

Automated user provisioning based on updates from external systems.
Streamlined user lifecycle management (create, update, delete).

**Comment by adlerhurst** [comment]
Additions to OP

## Acceptance Criteria
- [ ] The following **REST only** operations are implemented:

## User schema
- [ ] add schema extension which allows to select the schema type by jsonpath filter (e.g. exists urn:ietf:params:scim:schemas:extension:CustomExtensionName:2.0:User)
- [ ] add field extension for mapping scim field to schema field by jsonpath

**Comment by mffap** [comment]
Use case from a conversation: Delete user per SCIM is important when offering a "per-seat" pricing of a SaaS application. When a customer deletes/offboards an account the user should also be deleted from the IDP.

**Comment by sueess** [comment]
Today, we had another potential B2B SaaS customer who specifically requested SCIM. They consider other methods for user synchronization outdated and have listed SCIM as a requirement on their checklist for evaluating new software. Conclusion: We would highly appreciate the possibility :).

**Comment by hyperknot** [comment]
Just wanted to confirm from what I've researched. The inbound use case means that ZITADEL acts as a SCIM server and IdP platforms act as SCIM clients. ZITADEL provides the REST endpoints and the IdP admins click on their UI for adding/removing users to a given app. As a SaaS developer, inbound direction / SCIM server is what I need. Is that correct?

**Comment by hifabienne** [comment]
> Just want to confirm what I've researched. The inbound use case means that ZITADEL acts as a SCIM server and IdP platforms act as SCIM clients. ZITADEL provides the REST endpoints and the IdP admins / HR click on their UI for adding/removing users to a given app. As a SaaS developer, inbound direction / SCIM server is what I need. Is that correct?

You are right, inbound means an external service is able to provision users to ZITADEL through a SCIM interface.
Outbound on the other hand mean, that ZITADEL will send users out to an external service through a SCIM interface.
To make it more clear twi examples:
- Inbound: You create a user in your entra id, the user is then sent to ZITADELs SCIM interface and will be created in ZITADEL.
- Outbound: You create a user in ZITADEL, the user will then be sent to your entra ID with the SCIM standard, and can be created in your entra ID.

**Comment by hifabienne** [comment]
@muhlemmer @livio-a @stebenz I have put together the mapping from SCIM to ZITADEL.
Most attributes should not be a problem. 
Concerns from my side where we need to look into is:
- ZITADEL: loginname, email verified?
- SCIM: Email and phone is a list, ZITADEL has only one --> I would map the primary email from scim

let me know your thoughts

## Human User Mapping

SCIM | ZITADEL
-- | --
username | username
  | loginnames
name |  
name.formatted | profile.displayName
name.familyName | profile.familyName
name.givenName | profile.givenName
name.middleName | -
name.honorificPrefix | -
name.honorificSuffix | -
displayName | profile.displayName
nickName | profile.nickName
 | gender
profileUrl |  
title |  
preferredLanguage | profile.preferredLanguage
locale |  
timezone |  
active | user.state
password | password.password
Emails (list) </br>- Value </br>- Type </br>- primary | Email (single)</br>- email.emaile </br>- mail.isVerified?
phoneNumbers  (list) </br>- Value </br>- type | Phone  (single) </br>- phone.phone </br>- phone.isVerified
ims | -
photos | -
addresses | -
groups | -
entitlements | -
roles | -

## Service User Mapping

SCIM | ZITADEL
-- | --
username | username
  | loginnames
name | name
name.formatted | -
name.familyName | -
name.givenName | -
name.middleName | -
name.honorificPrefix | -
name.honorificSuffix | -
displayName | profile.displayName
nickName | profile.nickName
 | description
profileUrl |  
title |  
preferredLanguage |  
locale |  
timezone |  
active | user.state
password |  
Emails (list)  </br>- Value </br>- Type </br>-  primary |  
phoneNumbers  (list) </br>- Value </br>- type |  
ims | -
photos | -
addresses | -
groups | -
entitlements | -
roles | -

</div></b>

**Comment by livio-a** [comment]
A remark to the mapping:
Shouldn't the `externalIid` be mapped to a IdP link on the user?
Additionally, (maybe only for the future) wouldn't it make sense to have some customization on mapping groups to authorizations before we have the [groups feature](https://github.com/zitadel/zitadel/issues/5822).

Typical use case i see is that one wants to onboard a new employee and provision their accounts from EntraID to ZITADEL and grant necessary permissions. As soon as the user then first signs in, ZITADEL ideally already know the federated userid and does not even need to do JIT provisioning / linking.

**Comment by hifabienne** [comment]
> A remark to the mapping: Shouldn't the `externalIid` be mapped to a IdP link on the user? Additionally, (maybe only for the future) wouldn't it make sense to have some customization on mapping groups to authorizations before we have the [groups feature](https://github.com/zitadel/zitadel/issues/5822).
> 
> Typical use case i see is that one wants to onboard a new employee and provision their accounts from EntraID to ZITADEL and grant necessary permissions. As soon as the user then first signs in, ZITADEL ideally already know the federated userid and does not even need to do JIT provisioning / linking.

I agree on both, the externalid might be difficult as we would also need to know to which idp.
For a first mvp I would leave that out. and we can extend in a second version

**Comment by muhlemmer** [comment]
I think we can leave the org ID logically the same as the other APIs. As the SCIM requests are authenticated by a machine user, we can default to it's organization. We also have the org header to override this.

In any case the machine user needs proper permissions to manage users on a given org.

Therefore I think we should reuse the existing middleware and definition of org selection. We can go with custom fields later if needed.

**Comment by hifabienne** [comment]
> I think we can leave the org ID logically the same as the other APIs. As the SCIM requests are authenticated by a machine user, we can default to it's organization. We also have the org header to override this.
> 
> In any case the machine user needs proper permissions to manage users on a given org.
> 
> Therefore I think we should reuse the existing middleware and definition of org selection. We can go with custom fields later if needed.

As SCIM is a standard implementation, I am pretty sure those SCIM clients can't just send an organization header.
A custom field would be the better choice in that case. I guess just adding the org id as custom field would already be enough, without any other email domain logic.

**Comment by muhlemmer** [comment]
Looks good to me. I edited some metadata keys as the `profileURL` key was used in multiple places. Possibly copy/paste error.

We should add that the Email and Phone `verified` option is configurable in ZITADEL, right? So that is is either true or false for all users.

**Comment by hifabienne** [comment]
> Looks good to me. I edited some metadata keys as the `profileURL` key was used in multiple places. Possibly copy/paste error.
> 
> We should add that the Email and Phone `verified` option is configurable in ZITADEL, right? So that is is either true or false for all users.

Great thanks, I added your recommendation and also the org id as custom attribute

**Comment by hifabienne** [comment]
## Proposal for PATCH:


SCIM | ZITADEL | PATCH
-- | -- | --
  id | userid | patch not possible
  externalid | user metadata: key: externalid | Add, Replace, and Remove
username | username | replace
  | loginnames |
name |  | 
name.formatted | profile.displayName | Add, Replace, and Remove
name.familyName | profile.familyName  | Add, Replace
name.givenName | profile.givenName  | Add, Replace
name.middleName | user metadata: key: name.middleName  | Add, Replace, and Remove
name.honorificPrefix | user metadata: key: name.honorificPrefix  | Add, Replace, and Remove
name.honorificSuffix | user metadata: key: name.honorificSuffix  | Add, Replace, and Remove
displayName | profile.displayName  | Add, Replace, and Remove
nickName | profile.nickName  | Add, Replace, and Remove
not mapped | gender | patch not possible
profileUrl |  user metadata: key: profileURL | Add, Replace, and Remove
title |  user metadata: key: title | Add, Replace, and Remove
preferredLanguage | profile.preferredLanguage | Add, Replace, and Remove
locale |  user metadata: key: locale  | Add, Replace, and Remove
timezone |  user metadata: key: timezone  | Add, Replace, and Remove
active | user.state | Replace (active = activate, active false = deactivate, only if it has changed)
password | password.password  | Add, Replace
Emails (list) </br>- Value </br>- Type </br>- primary | Email (single)</br> Note: only primary email will be mapped</br>- email.emaile </br>- mail.isVerified?   | Replace
phoneNumbers  (list) </br>- Value </br>- type | Phone  (single) </br> Note: only primary phone will be mapped</br>- phone.phone </br>- phone.isVerified  | Add, Replace, Remove
ims | user metadata: key: ims  | Add, Replace, Remove
photos | user metadata: key: photos  | Add, Replace, Remove
addresses | user metadata: key: addresses  | Add, Replace, Remove
groups | user metadata: key: groups  | Add, Replace, Remove
entitlements | user metadata: key: entitlements  | Add, Replace, Remove
roles | user metadata: key: roles  | Add, Replace, Remove
custom attribute: organiztationId | organizationId  |  patch not possible
